test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingFailureTest.java

```diff
@@ -0,0 +1,181 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.test.sai;
+
+import java.io.IOException;
+import java.util.concurrent.Callable;
+import java.util.function.BiConsumer;
+
+import net.bytebuddy.ByteBuddy;
+import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
+import net.bytebuddy.implementation.MethodDelegation;
+import net.bytebuddy.implementation.bind.annotation.SuperCall;
+import org.junit.Test;
+
+import org.apache.cassandra.distributed.Cluster;
+import org.apache.cassandra.distributed.api.IInvokableInstance;
+import org.apache.cassandra.distributed.api.NodeToolResult;
+import org.apache.cassandra.distributed.test.TestBaseImpl;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.disk.v1.SAICodecUtils;
+import org.apache.cassandra.index.sai.disk.v1.segment.SegmentBuilder;
+import org.apache.cassandra.index.sai.disk.v1.segment.SegmentMetadata;
+import org.apache.cassandra.utils.Throwables;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.store.IndexInput;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import static org.apache.cassandra.distributed.api.Feature.GOSSIP;
+import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+
+public class IndexStreamingFailureTest extends TestBaseImpl
+{
+    public static final String TEST_ERROR_MESSAGE = "Injected failure!";
+
+    @Test
+    public void testAvailabilityAfterFailedNonEntireFileStreaming() throws Exception
+    {
+        testAvailabilityAfterStreaming(ByteBuddyHelper::installFlushError, false);
+    }
+
+    @Test
+    public void testAvailabilityAfterFailedEntireFileStreaming() throws Exception
+    {
+        testAvailabilityAfterStreaming(ByteBuddyHelper::installValidationError, true);
+    }
+
+    private void testAvailabilityAfterStreaming(BiConsumer<ClassLoader, Integer> injectedError, boolean streamEntireSSTables) throws Exception
+    {
+        try (Cluster cluster = init(Cluster.build(2)
+                                           .withConfig(c -> c.with(NETWORK, GOSSIP).set("stream_entire_sstables", streamEntireSSTables))
+                                           .withInstanceInitializer(injectedError)
+                                           .start()))
```

Comment:
TODO: This test might be a little on the expensive side, so I might try to have the two cases share the same cluster. It should be possible w/ a couple flags...

Reply:
Sharing the cluster would be great for CI costs and risk of timing out.